### PR TITLE
Select correct item using keys when filtered (#29)

### PR DESCRIPTION
**Pull request by fmechant**
Filtering items when looking for which item is focused, so that the correct item is selected when using the arrow keys and enter key.

### CODE REVIEW

File: src/Dropdown.elm
Comment (PedroHLC) [inline src/Dropdown.elm:544]: I would avoid the code de-dup here as the Elm compiler can reduce the `(state.filterText |> String.toLower)` to its final value.

File: src/Dropdown.elm
Comment (PedroHLC) [inline src/Dropdown.elm:443]: This entire `update` function needs a refactoring now that I'm looking at it.

**Review by PedroHLC — COMMENTED** [review]
Your solution works, but it's just that I don't trust elm 0.19 optimizations very much.

Could you try this one instead?: <details>
```diff
diff --git a/src/Dropdown.elm b/src/Dropdown.elm
index 632e638..c6881cf 100644
--- a/src/Dropdown.elm
+++ b/src/Dropdown.elm
@@ -358,111 +358,126 @@ withListAttributes attrs (Config config) =
 
 -}
 update : Config item msg model -> Msg item -> model -> State item -> ( State item, Cmd msg )
-update (Config config) msg model (State state) =
+update (Config config) msg model ((State state) as untouchedState) =
+    case msg of
+        NoOp ->
+            ( untouchedState, Cmd.none )
+
+        OnBlur ->
+            ( State { state | isOpen = closeOnlyIfNotMultiSelect config state }, Cmd.none )
+
+        OnClickOutside ->
+            ( State { state | isOpen = False }, Cmd.none )
+
+        OnClickPrompt ->
+            let
+                isOpen =
+                    not state.isOpen
+
+                cmd =
+                    if isOpen then
+                        Task.attempt (\_ -> NoOp) (Dom.focus (state.id ++ "input-search"))
+
+                    else
+                        Cmd.none
+            in
+            ( State { state | isOpen = isOpen, focusedIndex = 0, filterText = "" }, Cmd.map config.dropdownMsg cmd )
+
+        OnSelect item ->
+            let
+                selectedItemsNew =
+                    selectedItemsAsList config model
+                        |> modifySelectedItems config.dropdownType item
+            in
+            ( State
+                { state
+                    | isOpen = closeOnlyIfNotMultiSelect config state
+                }
+            , updateSelectedItemsCommand config.onSelectMsg selectedItemsNew
+            )
+
+        OnFilterTyped val ->
+            ( State { state | filterText = val }, Cmd.none )
+
+        OnKeyDown key ->
+            updateKeyDown config key model state
+
+
+updateKeyDown : InternalConfig item msg model -> Key -> model -> InternalState -> ( State item, Cmd msg )
+updateKeyDown config key model state =
     let
         items =
             config.itemsFromModel model
 
-        selectedItems =
-            selectedItemsAsList config model
+        newIndex =
+            case key of
+                ArrowUp ->
+                    if state.focusedIndex > 0 then
+                        state.focusedIndex - 1
 
-        ( newState, newCommand ) =
-            case msg of
-                NoOp ->
-                    ( state, Cmd.none )
-
-                OnBlur ->
-                    ( { state | isOpen = closeOnlyIfNotMultiSelect config state }, Cmd.none )
-
-                OnClickOutside ->
-                    ( { state | isOpen = False }, Cmd.none )
-
-                OnClickPrompt ->
-                    let
-                        isOpen =
-                            not state.isOpen
-
-                        cmd =
-                            if isOpen then
-                                Task.attempt (\_ -> NoOp) (Dom.focus (state.id ++ "input-search"))
-
-                            else
-                                Cmd.none
-                    in
-                    ( { state | isOpen = isOpen, focusedIndex = 0, filterText = "" }, Cmd.map config.dropdownMsg cmd )
-
-                OnSelect item ->
-                    let
-                        selectedItemsNew =
-                            selectedItems
-                                |> modifySelectedItems config.dropdownType item
-                    in
-                    ( { state
-                        | isOpen = closeOnlyIfNotMultiSelect config state
-                      }
-                    , updateSelectedItemsCommand config.onSelectMsg selectedItemsNew
-                    )
-
-                OnFilterTyped val ->
-                    ( { state | filterText = val }, Cmd.none )
-
-                OnKeyDown key ->
-                    let
-                        newIndex =
-                            case key of
-                                ArrowUp ->
-                                    if state.focusedIndex > 0 then
-                                        state.focusedIndex - 1
-
-                                    else
-                                        0
-
-                                ArrowDown ->
-                                    if state.focusedIndex < List.length items - 1 then
-                                        state.focusedIndex + 1
-
-                                    else
-                                        List.length items - 1
-
-                                _ ->
-                                    state.focusedIndex
-
-                        isOpen =
-                            case key of
-                                Esc ->
-                                    False
-
-                                Enter ->
-                                    False
-
-                                _ ->
-                                    True
-
-                        maybeFocusedItem =
-                            items
-                                |> List.indexedMap (\i item -> ( i, item ))
-                                |> List.filter (\( i, _ ) -> i == state.focusedIndex)
-                                |> List.head
-                                |> Maybe.map Tuple.second
-
-                        cmd =
-                            case key of
-                                Enter ->
-                                    case maybeFocusedItem of
-                                        Just focusedItem ->
-                                            selectedItems
-                                                |> modifySelectedItems config.dropdownType focusedItem
-                                                |> updateSelectedItemsCommand config.onSelectMsg
-
-                                        Nothing ->
-                                            Cmd.none
-
-                                _ ->
-                                    Cmd.none
-                    in
-                    ( { state | focusedIndex = newIndex, isOpen = isOpen }, cmd )
+                    else
+                        0
+
+                ArrowDown ->
+                    if state.focusedIndex < List.length items - 1 then
+                        state.focusedIndex + 1
+
+                    else
+                        List.length items - 1
+
+                _ ->
+                    state.focusedIndex
+
+        isOpen =
+            case key of
+                Esc ->
+                    False
+
+                Enter ->
+                    False
+
+                _ ->
+                    True
+
+        loweredFilter =
+            String.toLower state.filterText
+
+        maybeFocusedItemFold list i =
+            case list of
+                item :: tail ->
+                    if String.contains loweredFilter (item |> config.itemToText |> String.toLower) then
+                        if i == state.focusedIndex then
+                            Just item
+
+                        else
+                            maybeFocusedItemFold tail <| i + 1
+
+                    else
+                        maybeFocusedItemFold tail i
+
+                [] ->
+                    Nothing
+
+        maybeFocusedItem =
+            maybeFocusedItemFold
+                items
+                0
     in
-    ( State newState, newCommand )
+    ( State { state | focusedIndex = newIndex, isOpen = isOpen }
+    , case key of
+        Enter ->
+            case maybeFocusedItem of
+                Just focusedItem ->
+                    selectedItemsAsList config model
+                        |> modifySelectedItems config.dropdownType focusedItem
+                        |> updateSelectedItemsCommand config.onSelectMsg
+
+                Nothing ->
+                    Cmd.none
+
+        _ ->
+            Cmd.none
+    )
 
 
 
```
</details>

This patch should solve this in `O(n)` in the update, avoid unnecessary processing for other messages (though Elm is probably doing this same optimization), trigger less when using the state in lazy, and hopefully make this `update` more readable.

The correct approach would be to cache the filtered items the same way Evan caches the sorted rows in his table example, but that's not for me to decide right now...

**Comment by fmechant** [comment]
Hi Pedro,

Thank you for the review.

Although I think you are optimizing prematurely, I do agree with some of your changes.

But, doing these changes increases the scope of this fix, and I am not sure I like that.

Maybe others can share their thoughts about this?

**Comment by fmechant** [comment]
I simplified the updateKeyDown.
Not sure if that is OK. Please let me know.

**Comment by fmechant** [comment]
I want to add a unit test for findFocusedItem, and even for update with a OnKeyDown message, but there are no tests, at the moment.
Do you want me to add these tests?

Disadvantage of testing findFocusedItem directly, is that it needs to be exposed. I guess we do not want that?

**Comment by fmechant** [comment]
If we also do not want to expose Msg(..), than testing the update becomes impossible, I think.

**Review by PedroHLC — APPROVED** [review]
I found your solution very elegant @fmechant 

> Disadvantage of testing findFocusedItem directly, is that it needs to be exposed. I guess we do not want that?

Yeah, maybe in the future we can add `elm-program-test` and simulate the key pressing and test it E2E instead.

You got my approval, but the final word of this package is from @jouderianjr .

**Comment by fmechant** [comment]
I do not have experience with `elm-program-test`, but I am willing to spend some time implementing it for using the keys.
Sounds like an interesting idea, no?

**Review by renanpvaz — APPROVED** [review]

**Comment by PedroHLC** [comment]
> I do not have experience with `elm-program-test`, but I am willing to spend some time implementing it for using the keys.
> Sounds like an interesting idea, no?

Yeah, this is really interesting. But, one thing I noticed is that we're missing the ARIA roles, and that would be essential to finding the proper element "to click" in the E2E tests. [This example has an adequate implementation of it](https://w3c.github.io/aria-practices/examples/listbox/listbox-collapsible.html).

If you want, feel free to open another issue and PR to track it, I'll add it to our CI for proper checking on every merge.

Just a warning, porting to `elm-program-test` is a little more complicated than it looks. But it's good to learn it for someone who works (or wants to work) with Elm.

File: src/Dropdown.elm
Comment (jouderianjr) [inline src/Dropdown.elm:419]: I think you reversed the operation here and in the `ArrowUp`, so the `ArrowDown` should move the cursor forward in the list

**Review by jouderianjr — CHANGES_REQUESTED** [review]

File: src/Dropdown.elm
Comment (PedroHLC) [reply src/Dropdown.elm:419]: Also, we want to open (`isOpen = True`) when `ArrowDown` or `ArrowUp`, otherwise we break navigation without the mouse. The same for `Enter`:
```elm
            if state.isOpen then
                -- What yours does...
                
            else
                ( State { state | isOpen = True }, Cmd.none )
```

**Review by PedroHLC — CHANGES_REQUESTED** [review]

File: src/Dropdown.elm
Comment (fmechant) [reply src/Dropdown.elm:419]: I switched the operations, indeed. Will fix it.

Comment (fmechant) [reply src/Dropdown.elm]: I am confused, @PedroHLC . What is the behavior we want?

Comment (fmechant) [reply src/Dropdown.elm]: Do I understand correctly that, if you press ArrowUp, ArrowDown, or Enter when the state.isOpen == False, you just want to set the state.isOpen to True?

Comment (fmechant) [reply src/Dropdown.elm]: I will let the code talk for itself. There you can see how I understood you want it to work.

File: src/Dropdown.elm
Comment (PedroHLC) [inline src/Dropdown.elm:425]: yeah, that's it, you can just simplify by removing the "not" and inverting the branches.

Comment (fmechant) [reply src/Dropdown.elm:425]: In an `if then else`, I prefer to put the short block right after the `if`, so for the `else`, you do not have to scroll up to find what the `else` is about.
If I use `if state.isOpen then`, we get a longer block, and the `else` is all alone by itself, if you know what I mean.

Generally thought, I avoid booleans altogether and use case statements. Here, that would mean that we would have a state.visibility which could be Hidden or Visible. Then it would become
`case state.visibility of
  Hidden -> ...
  Visible -> ...`
I am not sure if you want me to go this far.

**Review by jouderianjr — APPROVED** [review]

**Review by PedroHLC — APPROVED** [review]

**Pull request merged**
